client/lib/MerkleMineGenerator.js

```diff
@@ -40,29 +35,34 @@ module.exports = class MerkleMineGenerator {
         }
 
         console.log(`Validated locally generated Merkle root ${localRoot} with Merkle root stored in MerkleMine contract!`)
-    }
 
-    async checkStarted() {
-        const merkleMine = await this.getMerkleMine()
+        // Validate that the MerkleMine contract is in a started state
         const started = await merkleMine.methods.started().call()
 
         if (!started) {
             throw new Error(`Generation period has not started for MerkleMine contract`)
         }
-    }
 
-    async checkGenerated() {
-        const merkleMine = await this.getMerkleMine()
+        // Validate MerkleMine contract balance is sufficient for the token allocation generation
+        const tokensPerAllocation = await merkleMine.methods.tokensPerAllocation().call()
+        const token = await this.getToken()
+        const merkleMineBalance = await token.methods.balanceOf(this.merkleMineAddress).call()
+
+        if (merkleMineBalance < tokensPerAllocation) {
+            throw new Error(`Tokens per allocation is ${tokensPerAllocation.toString()} but MerkleMine contract only has balance of ${merkleMineBalance.toString()}`)
+        }
+
+        // Validate that the recipient's token allocation has not been generated
         const generated = await merkleMine.methods.generated(this.recipientAddress).call()
 
         if (generated) {
-            throw new Error(`Allocation for ${this.recipientAddress} already mined!`)
+            throw new Error(`Allocation for ${this.recipientAddress} already generated!`)
         }
```

Comment:
Just a thought - another check is to actually do the proof verification here to make sure it'll pass in the contract.